### PR TITLE
[Android] Add `onUserLeaveHint` support to ReactActivityDelegate

### PR DESCRIPTION
**Pull request by behenate**
## Summary:

This PR adds `onUserLeaveHint` support into the `ReactActivityDelegate`. It allows modules to receive an event every time user moves the app into the background. This is slightly different than `onPause` - it's called only when the user intentionally moves the app into the background, e.g. when receiving a call `onPause` should be called but `onUserLeaveHint` shouldn't.

This feature is especially useful for libraries implementing features like Picture in Picture (PiP), where using `onUserLeaveHint` is the [recommended way of auto-entering PiP](https://developer.android.com/develop/ui/views/picture-in-picture#:~:text=You%20might%20want%20to%20include%20logic%20that%20switches%20an%20activity%20into%20PiP%20mode%20instead%20of%20going%20into%20the%20background.%20For%20example%2C%20Google%20Maps%20switches%20to%20PiP%20mode%20if%20the%20user%20presses%20the%20home%20or%20recents%20button%20while%20the%20app%20is%20navigating.%20You%20can%20catch%20this%20case%20by%20overriding%20onUserLeaveHint()%3A) for android < 12.

## Changelog:

[ANDROID] [ADDED] - Added `onUserLeaveHint` support into `ReactActivityDelegate`

## Test Plan:

Tested in the `rn-tester` app - callbacks are correctly called on both old and new architecture.

### CODE REVIEW

File: packages/react-native/ReactAndroid/src/main/java/com/facebook/react/ReactHost.kt [hunk 1]
```diff
@@ -57,6 +57,9 @@ interface ReactHost {
   /** To be called when the host activity is resumed. */
   fun onHostResume(activity: Activity?)
 
+  /** To be called when the host activity is about to go into the background as the result of user choice. */
+  fun onUserLeaveHint(activity: Activity?)
```
Comment (cortinico) [inline packages/react-native/ReactAndroid/src/main/java/com/facebook/react/ReactHost.kt:61]: Please rename to
```suggestion
  fun onHostLeaveHint(activity: Activity?)
```

**Review by cortinico — CHANGES_REQUESTED** [review]

**Comment by cortinico** [comment]
This needs to be reverted as it's introducing a crash internally. I'll follow up on it tomorrow @behenate

File: packages/react-native/ReactAndroid/src/main/java/com/facebook/react/ReactInstanceManager.java [hunk 2]
```diff
@@ -585,6 +585,38 @@ public void onHostPause() {
...
+
+    if (mCurrentActivity != null) {
+      Assertions.assertCondition(
+        activity == mCurrentActivity,
+        "Called onUserLeaveHint on an activity that is not the current activity, this is incorrect! "
+          + "Current activity: "
+          + mCurrentActivity.getClass().getSimpleName()
+          + " "
+          + "Leaving activity: "
+          + activity.getClass().getSimpleName());
```
Comment (cortinico) [inline packages/react-native/ReactAndroid/src/main/java/com/facebook/react/ReactInstanceManager.java:609]: @behenate Ok so one of our product was crashing on this assert
I haven't had the time to investigate what's going on, but practically we should understand if this assert is actually necessary.

That's the stacktrace:

```
02-01 14:38:52.320  4500  4522 V fb-android-monitor: java.lang.AssertionError
02-01 14:38:52.320  4500  4522 V fb-android-monitor: 	at com.facebook.infer.annotation.Assertions.assertCondition(Assertions.java:82)
02-01 14:38:52.320  4500  4522 V fb-android-monitor: 	at com.facebook.react.ReactInstanceManager.onUserLeaveHint(ReactInstanceManager.java:598)
02-01 14:38:52.320  4500  4522 V fb-android-monitor: 	at com.facebook.react.ReactDelegate.onUserLeaveHint(ReactDelegate.java:107)
02-01 14:38:52.320  4500  4522 V fb-android-monitor: 	at com.facebook.react.ReactActivityDelegate.onUserLeaveHint(ReactActivityDelegate.java:127)
02-01 14:38:52.320  4500  4522 V fb-android-monitor: 	at com.facebook.react.ReactActivity.onUserLeaveHint(ReactActivity.java:111)
02-01 14:38:52.320  4500  4522 V fb-android-monitor: 	at android.app.Activity.performUserLeaving(Activity.java:8189)
02-01 14:38:52.320  4500  4522 V fb-android-monitor: 	at android.app.Instrumentation.callActivityOnUserLeaving(Instrumentation.java:1520)
02-01 14:38:52.320  4500  4522 V fb-android-monitor: 	at android.app.ActivityThread.performUserLeavingActivity(ActivityThread.java:4659)
02-01 14:38:52.320  4500  4522 V fb-android-monitor: 	at android.app.ActivityThread.handlePauseActivity(ActivityThread.java:4643)
02-01 14:38:52.320  4500  4522 V fb-android-monitor: 	at android.app.servertransaction.PauseActivityItem.execute(PauseActivityItem.java:46)
02-01 14:38:52.320  4500  4522 V fb-android-monitor: 	at android.app.servertransaction.TransactionExecutor.executeLifecycleState(TransactionExecutor.java:176)
02-01 14:38:52.320  4500  4522 V fb-android-monitor: 	at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:97)
02-01 14:38:52.320  4500  4522 V fb-android-monitor: 	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2066)
02-01 14:38:52.320  4500  4522 V fb-android-monitor: 	at android.os.Handler.dispatchMessage(Handler.java:106)
02-01 14:38:52.320  4500  4522 V fb-android-monitor: 	at android.os.Looper.loop(Looper.java:223)
02-01 14:38:52.320  4500  4522 V fb-android-monitor: 	at android.app.ActivityThread.main(ActivityThread.java:7656)
02-01 14:38:52.320  4500  4522 V fb-android-monitor: 	at java.lang.reflect.Method.invoke(Native Method)
02-01 14:38:52.320  4500  4522 V fb-android-monitor: 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
02-01 14:38:52.320  4500  4522 V fb-android-monitor: 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947)
02-01 14:38:52.320  4500  4522 V fb-android-monitor: 	at Z.init(unknown)
```

Could you resubmit 3b6c522 without this assert?

Comment (cortinico) [reply packages/react-native/ReactAndroid/src/main/java/com/facebook/react/ReactInstanceManager.java:609]: @behenate are you able to follow-up here?

Comment (behenate) [reply packages/react-native/ReactAndroid/src/main/java/com/facebook/react/ReactInstanceManager.java:609]: @cortinico Sorry for no reply, I'm currently out of office until February 14th, I will follow up when I'm back :)

Comment (cortinico) [reply packages/react-native/ReactAndroid/src/main/java/com/facebook/react/ReactInstanceManager.java:609]: Sure no rush 👍

**Pull request closed without merging**